src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -512,23 +512,27 @@ impl crate::CommandRunner for CommandRunner {
     };
 
     if result.exit_code == 0 && self.cache_write {
-      let write_result = self
-        .update_action_cache(
-          &context,
-          &request,
-          &result,
-          &self.metadata,
-          &command,
-          action_digest,
-          command_digest,
-        )
-        .await;
-      if let Err(err) = write_result {
-        log::warn!("Failed to write to remote cache: {}", err);
-        context
-          .workunit_store
-          .increment_counter(Metric::RemoteCacheWriteErrors, 1);
-      };
+      let command_runner = self.clone();
+      let result = result.clone();
+      tokio::spawn(async move {
```

Comment:
I have a few concerns:

1. First, this code puts the upload into the background, but does not ensure that the Pants session blocks (for a reasonable amount of time) at the end to let the work complete. For example, with pantsd disabled, the Pants process could exit without even checking whether the uploads could still complete in a reasonable amount of time. Thus, the uploads will fail mid-stream because Pants exited. And this could be the predominant behavior.

When I originally envisioned async writes as an idea, the thought was to have a list of tasks in the Session that would block the Session from ending.

2. I'm also concerned about the attempt to increment the counter outside of an active workunit. How is that counter increment accounted for when there is no active workunit? (Or if it sees an active workunit, how is that increment handled when the workunit store has already declared the (parent?) workunit to be completed?)

Reply:
> First, this code puts the upload into the background, but does not ensure that the Pants session blocks (for a reasonable amount of time) at the end to let the work complete.

It isn't desirable to block for this to complete, either when `pantsd` is enabled (because it will proceed in the background), or when it is disabled (because then you are directly making a user's run slower).

> the Pants process could exit without even checking whether the uploads could still complete in a reasonable amount of time. Thus, the uploads will fail mid-stream because Pants exited. And this could be the predominant behavior.

We should not be biasing for the no-`pantsd` case. But if we were, to make the experience of blocking not suck, we'd want to fork the process to complete the write or something... which is non-trivial.

Reply:
> or when it is disabled (because then you are directly making a user's run slower).

I can imagine some scenarios where a user is fine with this tradeoff. For example, in CI, accepting a slight slowdown on the first run so that all future runs can leverage the remote cache.

> or require pantsd to be running to service cache writes, perhaps.

That sounds worth exploring. Would we change the implementation to more directly coordinate with panstsd? Or simply add validation to the options because we know things work much better when pantsd is used?

Reply:
> Would we change the implementation to more directly coordinate with panstsd?

This is what I was thinking. Something like "even if your run does not run 'in' `pantsd`, `pantsd` is launched and used for background work".

Reply:
Sounds good. And, to confirm, that is followup work to make this more robust, right? For now, we're only going to do `task_executor.spawn`?

Reply:
Correct. That and the metrics.